File: sdk/node/runtime-subjects.js
import { containsAllZeroPrincipal } from "./runtime-principals.js";

const RUNTIME_STATE_READ_SUBJECT_PATH = "runtime-state/read";

export function runtimeStateReadSubjectURA(realm, userID, errors) {
  const invalidInvocation = requiredErrorFactory(errors?.invalidInvocation, "invalidInvocation");
  const invalidRuntime = requiredErrorFactory(errors?.invalidRuntime, "invalidRuntime");
  const cleanRealm = runtimeStateSubjectSegment(realm, "realm", invalidInvocation);
  const cleanUserID = runtimeStateSubjectSegment(userID, "user_id", invalidInvocation);
  if (containsAllZeroPrincipal(cleanUserID)) {
    throw invalidInvocation("runtime-state read subject user_id must not be all-zero");
  }
  const subject = `easynet:///r/${cleanRealm}/resource/user.${cleanUserID}/${RUNTIME_STATE_READ_SUBJECT_PATH}`;
  if (!canonicalResourceSubject(subject)) {
    throw invalidRuntime("runtime-state read subject_ura must be canonical");
  }
  return subject;
}

export function isRuntimeStateReadSubjectURA(subjectURA) {
  const subject = canonicalResourceSubject(subjectURA);
  if (!subject || !subject.ownerID.startsWith("user.")) {
    return false;
  }
  const ownerUserID = subject.ownerID.slice("user.".length).trim();
  return (
    ownerUserID !== "" &&
    !containsAllZeroPrincipal(ownerUserID) &&
    subject.resourcePath === RUNTIME_STATE_READ_SUBJECT_PATH
  );
}

export function isRuntimeOwnerReadSubjectURA(subjectURA, calleeURA) {
  const subject = String(subjectURA ?? "").trim();
  const callee = String(calleeURA ?? "").trim();
  if (subject === "" || subject !== callee) {
    return false;
  }
  const subjectOwner = canonicalRuntimeOwnerSubject(subject);
  const calleeOwner = canonicalRuntimeOwnerSubject(callee);
  return (
    subjectOwner !== null &&
    calleeOwner !== null &&
    subjectOwner.kind === calleeOwner.kind &&
    subjectOwner.realm === calleeOwner.realm
  );
}

export function canonicalResourceSubject(subjectURA) {
  if (containsAllZeroPrincipal(subjectURA)) {
    return null;
  }
  const parsed = canonicalURAPath(subjectURA);
  if (!parsed || !parsed.path.startsWith("resource/")) {
    return null;
  }
  const resource = parsed.path.slice("resource/".length);
  const slash = resource.indexOf("/");
  if (slash <= 0 || slash === resource.length - 1) {
    return null;
  }
  const ownerID = resource.slice(0, slash).trim();
  const resourcePath = resource.slice(slash + 1).trim();
  if (
    ownerID === "" ||
    ownerID.includes("/") ||
    resourcePath === "" ||
    resourcePath.startsWith("/") ||
    resourcePath.includes("//")
  ) {
    return null;
  }
  return { ownerID, resourcePath };
}

function canonicalRuntimeOwnerSubject(subjectURA) {
  if (containsAllZeroPrincipal(subjectURA)) {
    return null;
  }
  const parsed = canonicalURAPath(subjectURA);
  if (!parsed) {
    return null;
  }
  if (parsed.path === "authority") {
    return { kind: "authority", realm: parsed.realm };
  }
  if (parsed.path.startsWith("device/")) {
    const deviceID = parsed.path.slice("device/".length).trim();
    if (deviceID !== "" && !deviceID.includes("/")) {
      return { kind: "device", realm: parsed.realm };
    }
  }
  if (parsed.path.startsWith("agent/device.")) {
    const scopedAgentID = parsed.path.slice("agent/device.".length).trim();
    const separator = scopedAgentID.indexOf(".");
    if (separator > 0 && separator < scopedAgentID.length - 1 && !scopedAgentID.includes("/")) {
      return { kind: "system-agent", realm: parsed.realm };
    }
  }
  return null;
}

function runtimeStateSubjectString(value, field, invalidInvocation) {
  if (typeof value !== "string" || value.trim() === "") {
    throw invalidInvocation(`runtime-state read subject ${field} is required`);
  }
  return value.trim();
}

function runtimeStateSubjectSegment(value, field, invalidInvocation) {
  const clean = runtimeStateSubjectString(value, field, invalidInvocation);
  if (clean.includes("/") || clean.includes("?") || clean.includes("#")) {
    throw invalidInvocation(`runtime-state read subject ${field} is not canonical`);
  }
  return clean;
}

function canonicalURAPath(value) {
  const raw = String(value ?? "").trim();
  const prefix = "easynet:///r/";
  if (!raw.startsWith(prefix)) {
    return null;
  }
  const rest = raw.slice(prefix.length);
  const slash = rest.indexOf("/");
  if (slash <= 0 || slash === rest.length - 1) {
    return null;
  }
  return {
    realm: rest.slice(0, slash),
    path: rest.slice(slash + 1),
  };
}

function requiredErrorFactory(factory, name) {
  if (typeof factory === "function") {
    return factory;
  }
  throw new Error(`runtime subject ${name} error factory is required`);
}

File: sdk/go/runtime_subjects.go
package easynet

import "strings"

const runtimeStateReadSubjectPath = "runtime-state/read"

// RuntimeStateReadSubjectURA builds the canonical subject for runtime-state
// read projections owned by one authenticated user.
//
// Products pass realm/user identity into this helper instead of defaulting
// read-only invocations to the target device. Admission still validates the
// resulting tuple; this helper only centralizes the product-neutral subject
// projection used by history, catalogue, and status reads.
func RuntimeStateReadSubjectURA(realm string, userID string) (string, error) {
	realm = strings.TrimSpace(realm)
	userID = strings.TrimSpace(userID)
	if realm == "" {
		return "", invalidInvocation("runtime-state read subject realm is required", nil)
	}
	if userID == "" {
		return "", invalidInvocation("runtime-state read subject user_id is required", nil)
	}
	if containsAllZeroPrincipal(userID) {
		return "", invalidInvocation("runtime-state read subject user_id must not be all-zero", nil)
	}
	subject := ResourceDotURA(realm, "user."+userID, runtimeStateReadSubjectPath)
	if _, err := ParseURA(subject); err != nil {
		return "", invalidInvocation("runtime-state read subject_ura must be canonical", err)
	}
	return subject, nil
}

func isRuntimeStateReadSubjectURA(subjectURA string) bool {
	parts, err := ParseURAParts(strings.TrimSpace(subjectURA))
	if err != nil || parts.Kind != URAKindResource {
		return false
	}
	ownerID := strings.TrimSpace(parts.OwnerID)
	userID := strings.TrimPrefix(ownerID, "user.")
	return strings.HasPrefix(ownerID, "user.") &&
		strings.TrimSpace(userID) != "" &&
		!containsAllZeroPrincipal(userID) &&
		strings.TrimSpace(parts.Path) == runtimeStateReadSubjectPath
}

func isRuntimeGovernanceReadSubjectURA(subjectURA string, calleeURA string) bool {
	subjectURA = strings.TrimSpace(subjectURA)
	calleeURA = strings.TrimSpace(calleeURA)
	if isRuntimeStateReadSubjectURA(subjectURA) {
		return true
	}
	subject, subjectErr := ParseURAParts(subjectURA)
	callee, calleeErr := ParseURAParts(calleeURA)
	if subjectErr != nil || calleeErr != nil {
		return false
	}
	return (subject.Kind == URAKindAuthority || subject.Kind == URAKindDevice) &&
		subject.Kind == callee.Kind &&
		subject.Realm == callee.Realm &&
		subjectURA == calleeURA
}

// RuntimeGovernanceReadSubjectURA projects a business subject into the
// canonical runtime governance-read subject accepted by provider-backed
// descriptor and receipt-history reads.
//
// Runtime governance reads are not product actions against the target device.
// User-, agent-, ability-, and user-owned resource subjects are projected to
// the user's runtime-state/read resource; a runtime owner subject is admitted
// only when it exactly matches the selected callee runtime owner.
func RuntimeGovernanceReadSubjectURA(subjectURA string, calleeURA string) (string, error) {
	subjectURA = strings.TrimSpace(subjectURA)
	calleeURA = strings.TrimSpace(calleeURA)
	if subjectURA == "" {
		return "", invalidInvocation("runtime governance read subject_ura is required", nil)
	}
	if isRuntimeGovernanceReadSubjectURA(subjectURA, calleeURA) {
		return subjectURA, nil
	}
	parts, err := ParseURAParts(subjectURA)
	if err != nil {
		return "", invalidInvocation("runtime governance read subject_ura must be canonical", err)
	}
	switch parts.Kind {
	case URAKindUser:
		return RuntimeStateReadSubjectURA(parts.Realm, parts.UserID)
	case URAKindAgent:
		if strings.TrimSpace(parts.UserID) == "" {
			break
		}
		return RuntimeStateReadSubjectURA(parts.Realm, parts.UserID)
	case URAKindAbility:
		userID := strings.TrimSpace(parts.AbilityOwner.UserID)
		if userID == "" {
			userID = strings.TrimSpace(parts.UserID)
		}
		if userID != "" {
			return RuntimeStateReadSubjectURA(parts.Realm, userID)
		}
	case URAKindResource:
		ownerID := strings.TrimSpace(parts.OwnerID)
		if strings.HasPrefix(ownerID, "user.") {
			userID := strings.TrimPrefix(ownerID, "user.")
			if strings.TrimSpace(userID) != "" && !strings.Contains(userID, ".") && !strings.Contains(userID, "/") {
				return RuntimeStateReadSubjectURA(parts.Realm, userID)
			}
		}
	}
	return "", invalidInvocation("runtime governance read subject_ura must be a runtime owner or user-owned runtime-state read subject", nil)
}

func runtimeGovernanceReadSubjectURA(subjectURA string, calleeURA string) (string, error) {
	return RuntimeGovernanceReadSubjectURA(subjectURA, calleeURA)
}
